### ISSUE TITLE
Only support when num_heads is even in transformer

### ISSUE THREAD

**Issue by EvanZ** [pytorch/pytorch#83355]
### 🐛 Describe the bug

I have verified that up until v.1.11 I can set `nhead=1` in `TransformerEncoderLayer`, but in v.1.12 this throws a Runtime Error:

```
Exception has occurred: RuntimeError
Only support when num_heads is even in transformer

```

As far as I can tell the docs do not specify that `nhead` must be even, unless I'm missing it. 

### Versions

```
Collecting environment information...
PyTorch version: 1.12.0+cu102
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.4 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.22.4
Libc version: glibc-2.31

Python version: 3.8.10 (default, Jun 22 2022, 20:18:18)  [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-5.13.0-1025-aws-x86_64-with-glibc2.29
Is CUDA available: True
CUDA runtime version: 11.0.221
GPU models and configuration: GPU 0: Tesla V100-SXM2-16GB
Nvidia driver version: 510.47.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.1
[pip3] pytorch-lightning==1.7.1
[pip3] torch==1.12.0
[pip3] torch-tb-profiler==0.4.0
[pip3] torchinfo==1.7.0
[pip3] torchmetrics==0.9.3
[pip3] torchtext==0.13.0
[conda] Could not collect
```

cc @ezyang @gchanan @zou3519 @ngimel @jbschlosser @bhosmer @cpuhrsch @erichan1

**Comment by erichan1** [comment]
Reproduced! We are fixing. Thanks @EvanZ

**Comment by ngimel** [comment]
It could be tricky to fix the masked softmax kernel, so probably the easiest thing is to send odd number of heads to the slow path.

**Comment by erichan1** [comment]
Yep we'll send to slow path. Root cause is here https://github.com/pytorch/pytorch/blob/b2363520363d8c85496f6954fde005834c188b5d/aten/src/ATen/native/cuda/SoftMax.cu#L1009. I would hope there aren't too many users of odd num MHA heads... but this is definitely a major issue.

**Comment by EvanZ** [comment]
To be clear, I really was looking just to maintain support for 1 head, not an odd number of heads generally. In my case testing showed that 2 heads was not better than 1 lol.

**Comment by ngimel** [comment]
Then probably you want to use 2 heads, because 2 heads would go to a fast code path, whereas one head will go to a slow one, so it'll probably end up being slower than 2.

**Comment by EvanZ** [comment]
In my testing 2 heads is also slower than using 1. I also am relying on a pretrained model with 1 head, so having to use 2 heads would be an issue there as well.

**Comment by erichan1** [comment]
https://github.com/pytorch/pytorch/pull/83483 Fixes and adds test
